File: solutions/code-carl/ch06_tree/t15_lc0112_has_path_sum.py
"""LeetCode 112. Path Sum
https://leetcode.cn/problems/path-sum/
"""

from typing import Optional

from tree_node import TreeNode


class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        if not root.left and not root.right:
            return root.val == targetSum
        if self.hasPathSum(root.left, targetSum - root.val):
            return True
        return self.hasPathSum(root.right, targetSum - root.val)


if __name__ == "__main__":
    solution = Solution()
    root1 = TreeNode(
        5,
        TreeNode(4, TreeNode(11, TreeNode(7), TreeNode(2))),
        TreeNode(8, TreeNode(13), TreeNode(4, None, TreeNode(1))),
    )
    assert solution.hasPathSum(root1, 22) is True

    root2 = TreeNode(1, TreeNode(2), TreeNode(3))
    assert solution.hasPathSum(root2, 5) is False

    assert solution.hasPathSum(None, 0) is False
    print("All tests passed.")

File: solutions/code-carl/ch06_tree/T15_LC0112_HasPathSum.java
/**
 * Practice template for:
 * - LeetCode 112. Path Sum
 *   https://leetcode.cn/problems/path-sum/
 */
public class T15_LC0112_HasPathSum {

    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }
        return backtracking(root, targetSum);
    }

    public boolean backtracking(TreeNode root, int sum) {
        if (root.left == null && root.right == null) {
            return sum == root.val;
        }

        if (root.left != null) {
            if (backtracking(root.left, sum - root.val)) {
                return true;
            }
        }
        if (root.right != null) {
            return backtracking(root.right, sum - root.val);
        }
        return false;
    }

}
